pkg/controller/daemon/daemon_controller.go

```diff
@@ -1376,3 +1399,85 @@ func getUnscheduledPodsWithoutNode(runningNodesList []*v1.Node, nodeToDaemonPods
 
 	return results
 }
+
+// runNodeUpdateWorker is a worker that processes node updates from the nodeUpdateQueue.
+func (dsc *DaemonSetsController) runNodeUpdateWorker(ctx context.Context) {
+	for dsc.processNextNodeUpdate(ctx) {
+	}
+}
+
+func (dsc *DaemonSetsController) processNextNodeUpdate(ctx context.Context) bool {
+	key, quit := dsc.nodeUpdateQueue.Get()
+	if quit {
+		return false
+	}
+	defer dsc.nodeUpdateQueue.Done(key)
+
+	err := dsc.syncNodeUpdate(ctx, key)
+	if err == nil {
+		dsc.nodeUpdateQueue.Forget(key)
+		return true
+	}
+
+	logger := klog.FromContext(ctx)
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	logger.Error(err, "Error syncing node update, requeuing", "node", key.nodeName)
+	dsc.nodeUpdateQueue.AddRateLimited(key)
+
+	return true
+}
+
+func (dsc *DaemonSetsController) syncNodeUpdate(ctx context.Context, item *NodeUpdateItem) error {
+	logger := klog.FromContext(ctx)
+
+	node, err := dsc.nodeLister.Get(item.nodeName)
+	if apierrors.IsNotFound(err) {
+		logger.V(3).Info("Node not found, skipping update", "node", item.nodeName)
+		return nil
+	}
+	if err != nil {
+		logger.V(4).Info("Error getting node", "node", item.nodeName, "err", err)
+		return err
+	}
+
+	dsList, err := dsc.dsLister.List(labels.Everything())
+	if err != nil {
+		logger.V(4).Info("Error listing daemon sets", "err", err)
+		return err
+	}
```

Comment:
we still keep listing, what is the benefit of adding this complexity?

Reply:
Thanks @aojea  for the review. When a Node event is detected, it is necessary to check which DaemonSets can be scheduled on that Node (this involves calling NodeShouldRunDaemonPod()). If the Node can be scheduled for a DaemonSet, the key of that DaemonSet is added to the DaemonSet workqueue.

Reply:
I see, so we build a two tiers architecture, something like this?


--> node event handler --> node workqueue --> get ds from node  ------|--> ds workqueue
--> ds event handler ----------------------------------------------------|

Reply:
Yes, this helps avoid inefficient node eventhandler processing, which could otherwise cause the NodeInformer's `pendingNotifications` to grow infinitely in memory, leading to an OOM.